### ISSUE TITLE
FP32 accumulation of bf16 gradients in FSDP

### ISSUE THREAD

**Issue by zplizzi** [pytorch/pytorch#106395]
### 🚀 The feature, motivation and pitch

I was training a model (1B and 7B param llama-like architecture) using FSDP in bf16, and found that it trained well on 12x8 GPUs, but the training would become unstable on 48x8 GPUs. After a good deal of debugging, we found that changing `reduce_dtype` to `None` instead of `bf16` resolved the issue. However, this fix isn't ideal, as the model now runs slower since it's sending around full-precision gradients. Admittedly we have a sorta strange setup, where we leave `param_dtype=None` and instead do mixed precision using the `torch.cuda.amp(torch.bfloat16)` wrapper. And we're also still using `torch==1.13`. 

Doing some research, it looks like [deepspeed has implemented](https://github.com/microsoft/DeepSpeed/issues/2352) bf16 gradient reduction with fp32 accumulation. So they are only transferring half-precision gradients, but accumulating/reducing them in full-precision, since it's well known that accumulation operations are often not super stable in half-precision. I was wondering if FSDP has implemented something similar to this, and if not, would request that it be implemented.

cc @zhaojuanmao @mrshenli @rohan-varma @awgu @fegin @penguinwu

**Comment by awgu** [comment]
My understanding is that there are two separate concerns:
1. The accumulation dtype used in NCCL reduce-scatter
2. The accumulation dtype used in FSDP if there is an existing sharded gradient from the previous backward

I think that FSDP does not have agency over the 1st. For the 2nd, we do already perform the accumulation (i.e. the sum) in fp32 (by upcasting the reduce-scattered gradient to fp32 before summing):
https://github.com/pytorch/pytorch/blob/6d86a255e6e4af39367b51903e234417b683221d/torch/distributed/fsdp/_runtime_utils.py#L809-L820

I need to look into the DeepSpeed PR more, but I wanted to mention the above first.

**Comment by zplizzi** [comment]
We looked at the deepspeed implementation more closely and it seems like I was mistaken, and it's just doing # 2 in mixed-precision (accumulation between steps) but not the inter-GPU accumulation. So it has the same behavior as FSDP.

Whether it's possible to do the inter-GPU accumulation in mixed precision I'm still unsure about, it might be impossible. Certainly it doesn't seem like NCCL exposes that option.
